third_party/move/move-compiler-v2/src/pipeline/lint_processor/needless_mutable_reference.rs

```diff
@@ -0,0 +1,236 @@
+// Copyright (c) Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+//! This module implements a stackless-bytecode linter that checks for mutable references
+//! that are never used mutably, and suggests to use immutable references instead.
+//! For example, if a mutable reference is never written to or passed as a mutable reference
+//! parameter to a function call, it can be replaced with an immutable reference.
+//!
+//! Currently, we only track mutable references that are:
+//! - function parameters,
+//! - obtained via `&mut` or `borrow_global_mut`.
+
+use crate::{lint_common::LintChecker, pipeline::lint_processor::StacklessBytecodeLinter};
+use move_model::{
+    ast::TempIndex,
+    model::{GlobalEnv, Loc, Parameter},
+};
+use move_stackless_bytecode::{
+    function_target::FunctionTarget,
+    stackless_bytecode::{Bytecode, Operation},
+};
+use std::collections::{BTreeMap, BTreeSet};
+
+/// Track "mutable" usages of certain mutable references in a function.
+/// Currently, the tracking is performed conservatively, in a flow-insensitive
+/// manner, to minimize perceived false positives.
+#[derive(Default)]
+struct MutableReferenceUsageTracker {
+    /// Keys are temps which are origins of certain mutable references.
+    /// Each key is mapped to a location where it acquires the mutable reference.
+    /// The origins tracked currently are:
+    /// - function parameters that are mutable references.
+    /// - mutable references acquired through `&mut` or `borrow_global_mut`.
+    /// The list above can be extended in the future.
+    origins: BTreeMap<TempIndex, Loc>,
+    /// Derived edges from mutable references.
+    /// A derived edge y -> x is created in the following cases:
+    /// - `x = y;`,  where y: &mut
```

Comment:
What about `x = f(y, z)` where y and/or z are mutable references?

Reply:
it is at bytecode level so not an issue I guess?

Reply:
Sorry, I am not sure I understand the question.

If we have `x = f(y, z);` such that `y: &mut u64`, then this (intra-procedural) checker assumes `y` has been mutably used and does not complain. It does not continue tracking `x` (note that we currently only track `&mut` and `borrow_global_mut`, as well as things that can derived from them through assignments and field borrows).

If we want to track arbitrary returns from the function, like in `x = f(y, z);`, it may not be possible to suggest a simple alternative (like change to immutable borrow/ref) to the user and will also lead to too many false positives.